src/Features/Core/Portable/Snippets/SnippetProviders/AbstractForLoopSnippetProvider.cs

```diff
@@ -11,9 +12,36 @@ namespace Microsoft.CodeAnalysis.Snippets.SnippetProviders
     internal abstract class AbstractForLoopSnippetProvider : AbstractInlineStatementSnippetProvider
     {
         protected override bool IsValidAccessingType(ITypeSymbol type, Compilation compilation)
-            => type.IsIntegralType() || type.IsNativeIntegerType;
+        {
+            if (IsSuitableIntegerType(type))
+            {
+                return true;
+            }
+
+            var hasLengthProperty = FindLengthProperty(type, compilation) is not null;
+            var hasCountProperty = FindCountProperty(type, compilation) is not null;
+
+            // We want to allow types, which have either `Length` or `Count` property, but not both to avoid ambiguity
+            return hasLengthProperty ^ hasCountProperty;
+        }
 
         protected override Func<SyntaxNode?, bool> GetSnippetContainerFunction(ISyntaxFacts syntaxFacts)
             => syntaxFacts.IsForStatement;
+
+        protected static bool IsSuitableIntegerType(ITypeSymbol type)
+            => type.IsIntegralType() || type.IsNativeIntegerType;
+
+        protected static IPropertySymbol? FindLengthProperty(ITypeSymbol type, Compilation compilation)
+            => FindAccessibleIntegerProperty(type, compilation, "Length");
+
+        protected static IPropertySymbol? FindCountProperty(ITypeSymbol type, Compilation compilation)
+            => FindAccessibleIntegerProperty(type, compilation, "Count");
+
+        private static IPropertySymbol? FindAccessibleIntegerProperty(ITypeSymbol type, Compilation compilation, string propertyName)
+        {
+            return type
+                .GetAccessibleMembersInThisAndBaseTypes<IPropertySymbol>(compilation.Assembly)
```

Comment:
is there a GetAccessibleMembers that takes a name?  would be a lot better than getting all members (Expensive) and having to iterate over all of them (expensive).